File: packages/cli/hieudoanm/cobra.go/src/cmd/gh/license/license.go
package license

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"

	"github.com/AlecAivazis/survey/v2"
	"github.com/hieudoanm/jack/src/cmd/gh/shared"
	requests "github.com/hieudoanm/jack/src/libs/requests"
	"github.com/spf13/cobra"
)

type license struct {
	Key    string `json:"key"`
	Name   string `json:"name"`
	SpdxID string `json:"spdx_id"`
	URL    string `json:"url"`
	Body   string `json:"body"`
}

func fetchLicense(spdxID, output string, fetch shared.FetchFunc) error {
	body, err := fetch("https://api.github.com/licenses", requests.Options{
		Header: http.Header{
			"Accept":     {"application/json"},
			"User-Agent": {"hieudoanm-cli"},
		},
	})
	if err != nil {
		return fmt.Errorf("error fetching licenses: %w", err)
	}

	var licenses []license
	if err := json.Unmarshal(body, &licenses); err != nil {
		return fmt.Errorf("error parsing response: %w", err)
	}

	selectedSpdxID := spdxID
	if selectedSpdxID == "" {
		var spdxIds []string
		for _, l := range licenses {
			spdxIds = append(spdxIds, l.SpdxID)
		}

		prompt := &survey.Select{
			Message: "Select a license:",
			Options: spdxIds,
		}
		if err := survey.AskOne(prompt, &selectedSpdxID); err != nil {
			return fmt.Errorf("selection cancelled: %w", err)
		}
	}

	url := fmt.Sprintf("https://api.github.com/licenses/%s", selectedSpdxID)
	detailBody, err := fetch(url, requests.Options{
		Header: http.Header{
			"Accept":     {"application/json"},
			"User-Agent": {"hieudoanm-cli"},
		},
	})
	if err != nil {
		return fmt.Errorf("error fetching license: %w", err)
	}

	var lic license
	if err := json.Unmarshal(detailBody, &lic); err != nil {
		return fmt.Errorf("error parsing response: %w", err)
	}

	file, err := os.Create(output)
	if err != nil {
		return fmt.Errorf("error creating file: %w", err)
	}
	defer file.Close()

	if _, err := file.WriteString(lic.Body); err != nil {
		return fmt.Errorf("error writing file: %w", err)
	}

	fmt.Printf("Written %s (%s) to %s\n", lic.Name, lic.SpdxID, output)
	return nil
}

func NewCommand() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "license",
		Short: "Fetch a license template from GitHub",
		Long: `Fetch and save a license template from the GitHub licenses API.

Fetches the list of available licenses, prompts the user to select
one (or uses --spdx-id), then writes the license body to a file.`,
		Example: `  gh license
  gh license --spdx-id MIT
  gh license --spdx-id Apache-2.0 -o LICENSE.txt`,
		RunE: func(cmd *cobra.Command, args []string) error {
			jsonOutput, _ := cmd.Flags().GetBool("json")
			spdxID, _ := cmd.Flags().GetString("spdx-id")
			output, _ := cmd.Flags().GetString("output")

			if jsonOutput {
				if spdxID == "" {
					return fmt.Errorf("--spdx-id is required for JSON output")
				}
				url := fmt.Sprintf("https://api.github.com/licenses/%s", spdxID)
				detailBody, err := shared.FetchFuncDefault(url, requests.Options{
					Header: http.Header{
						"Accept":     {"application/json"},
						"User-Agent": {"hieudoanm-cli"},
					},
				})
				if err != nil {
					return fmt.Errorf("error fetching license: %w", err)
				}
				var lic license
				if err := json.Unmarshal(detailBody, &lic); err != nil {
					return fmt.Errorf("error parsing response: %w", err)
				}
				out, _ := json.MarshalIndent(map[string]interface{}{
					"name":    lic.Name,
					"spdx_id": lic.SpdxID,
					"url":     lic.URL,
					"body":    lic.Body,
				}, "", "  ")
				fmt.Println(string(out))
				return nil
			}

			return fetchLicense(spdxID, output, shared.FetchFuncDefault)
		},
	}

	cmd.Flags().String("spdx-id", "", "SPDX license identifier (skip prompt)")
	cmd.Flags().StringP("output", "o", "LICENSE", "Output file path")
	return cmd
}

File: packages/cli/hieudoanm/clap.rs/src/cmd/gh/license.rs
use anyhow::Context;
use serde::Deserialize;

#[derive(Debug, Deserialize)]
struct License {
    key: String,
    name: String,
    spdx_id: String,
    url: String,
    body: String,
}

fn fetch_licenses() -> anyhow::Result<Vec<License>> {
    let client = reqwest::blocking::Client::new();
    let resp = client
        .get("https://api.github.com/licenses")
        .header("Accept", "application/json")
        .header("User-Agent", "hieudoanm-cli")
        .send()
        .context("error fetching licenses")?;
    let licenses: Vec<License> = resp.json().context("error parsing response")?;
    Ok(licenses)
}

fn fetch_license(spdx_id: &str) -> anyhow::Result<License> {
    let url = format!("https://api.github.com/licenses/{spdx_id}");
    let client = reqwest::blocking::Client::new();
    let resp = client
        .get(&url)
        .header("Accept", "application/json")
        .header("User-Agent", "hieudoanm-cli")
        .send()
        .context("error fetching license")?;
    let lic: License = resp.json().context("error parsing response")?;
    Ok(lic)
}

#[derive(clap::Args)]
pub struct Args {
    #[arg(long = "spdx-id", help = "SPDX license identifier (skip prompt)")]
    pub spdx_id: Option<String>,
    #[arg(
        short = 'o',
        long = "output",
        default_value = "LICENSE",
        help = "Output file path"
    )]
    pub output: String,
}

pub fn command() -> clap::Command {
    clap::Command::new("license")
        .about("Fetch a license template from GitHub")
        .arg(
            clap::Arg::new("spdx-id")
                .long("spdx-id")
                .help("SPDX license identifier (skip prompt)"),
        )
        .arg(
            clap::Arg::new("output")
                .short('o')
                .long("output")
                .help("Output file path")
                .default_value("LICENSE"),
        )
}

pub async fn run(matches: &Args) -> anyhow::Result<()> {
    let spdx_id = matches.spdx_id.as_ref();
    let output = &matches.output;

    let selected = if let Some(sid) = spdx_id {
        sid.clone()
    } else {
        let licenses = fetch_licenses()?;
        let spdx_ids: Vec<&str> = licenses.iter().map(|l| l.spdx_id.as_str()).collect();
        if spdx_ids.is_empty() {
            anyhow::bail!("no licenses available");
        }
        println!("Available licenses:");
        for (i, sid) in spdx_ids.iter().enumerate() {
            println!("  {}. {sid}", i + 1);
        }
        print!("Select license by SPDX ID: ");
        std::io::Write::flush(&mut std::io::stdout())?;
        let mut input = String::new();
        std::io::stdin().read_line(&mut input)?;
        input.trim().to_string()
    };

    let lic = fetch_license(&selected)?;
    std::fs::write(output, &lic.body).context("error writing file")?;
    println!("Written {} ({}) to {output}", lic.name, lic.spdx_id);
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_command_definition() {
        let cmd = command();
        assert!(!cmd.get_name().is_empty());
    }
}
